### DISCUSSION TITLE
How find the block in my code?

### DISCUSSION THREAD

**Discussion by iknowbr12** [Rapptz/discord.py#6541] [Q&A] [answered]
Hi! In last weeks, i am reciveing this errrs in my console;
```
WARNING:discord.gateway:Can't keep up, shard ID 1 websocket is 41.3s behind.
Shard ID 0 heartbeat blocked for more than 10 seconds.
```
(On all shards)

I am also receiving complaints from my users about a lot of slowness and delay.
I see bot "die" for 20 seconds or high time...

Well, I believe this is a block in my code, somewhere.
I do not believe that it is a library in this case, but some part of my code, somewhere in:

```
# Check function helper
def QuestionChecker(self, algorithm, reply, msg):
    if algorithm == "logic_one": 
        return msg.lower().strip(" ") == reply
    elif algorithm == "logic_two":
        return msg.lower().strip(" ") in reply
    elif algorithm == "requirements":
        get_list = msg.lower().strip(" ").replace(" e ",",").split(",")
        clear_list = [i.strip(" ") for i in get_list]
        final_list = list(filter(lambda i: not i in ["e",","], clear_list))
 
        final_list.sort()
        reply.sort()
 
        return final_list == reply
    else:
        print(f"[BOT] (ERROR in QuestionChecker) The response algorithm '{algorithm}' was not recognized.")
 
    return False
 
# Main function
q = self.GetQuestion() # Call mysql query
question_id, question_text, question_reply, question_reply_algorithm, question_embed_text, question_thumb_url, question_collaborator_id, question_collaborator_public = q 
reply = self.TreatResponse(question_reply_algorithm, question_reply)
#[...]
 
# The check function
def check(message): 
    if message.channel == ctx.channel and message.author.bot == False:
        return self.QuestionChecker(question_reply_algorithm, reply, message.content)
    else: return False
 
try:
    self.TaskTimer = asyncio.create_task(self.TimeLeftTimer(ctx))
    self.WaitMessage = self.bot.wait_for('message', check=check, timeout=25)
 
    message_hit = await self.WaitMessage 
    self.TaskTimer.cancel()
    self.WaitMessage = None
    self.WF_Stamp = None
    if self.finished:
        return
	#[...]
except asyncio.TimeoutError:
   	await self.TimeoutRound(ctx, question_reply)

```

In special:
```
message_hit = await self.WaitMessage 
check() function
```
This error started to happen a short time ago, with no code updates. Just more servers using the bot.

Useful information
- I am using UBUNTU;
- Python 3.8.2;
- Discord.py 1.5.1;
- My bot have 4,5k servers, with 5 shards, using autosharedbot;

Do you have any idea what is causing this and how can I resolve it?
All help will be most welcome.

**Comment by mesub7** [comment, marked as answer]
Are you using an async version of mysql? If you aren't then that is most likely the reason.

**Reply by iknowbr12** [reply]
I am using mysql.connector(https://pypi.org/project/mysql-connector/), not async.
Do you recommend an async library?
Is there a way for me to confirm the problem?

**Reply by lgaan** [reply]
MySQL is not really a recommended to use as a database. However, if you really want to use it then I'd recommend having a look at [aiomysql](https://github.com/aio-libs/aiomysql)

**Reply by iknowbr12** [reply]
Thanks. Which database would be recommended and why?

**Reply by lgaan** [reply]
The common SQL databases we recommend are SQLite and Postgres.
